apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/Numeric.cs

```diff
@@ -0,0 +1,663 @@
+{
+    // Note: These types are internal and used for implementing BigQueryNumeric and
+    // BigQueryBigNumeric only, which are thoroughly tested.
+    // If we ever need to make these types public, then we need to test them on its own.
+    internal sealed class NumericType
+    {
+        private sealed class Parameters
+        {
+            internal Parameters(BigInteger scaledMinValue, BigInteger scaledMaxValue, int scale)
+            {
+                GaxPreconditions.CheckArgumentRange(scale, nameof(scale), 0, PowersOf10.Length - 1);
+                GaxPreconditions.CheckState(scaledMinValue <= scaledMaxValue,
+                    $"{nameof(scaledMinValue)}={scaledMinValue} is greater than {nameof(scaledMaxValue)}={scaledMaxValue}.");
+
+                Scale = scale;
+                ScaledMinValue = scaledMinValue;
+                ScaledMaxValue = scaledMaxValue;
+            }
+
+            // These are scaled by the scaling factor.
+            // A value of 1 is really 0.000000001 if the scale is 9.
+            public BigInteger ScaledMinValue { get; }
+            public BigInteger ScaledMaxValue { get; }
+
+            public int Scale { get; }
+
+            public override bool Equals(object obj) =>
+                obj is Parameters def
+                && def.ScaledMinValue.Equals(ScaledMinValue)
+                && def.ScaledMaxValue.Equals(ScaledMaxValue)
+                && def.Scale.Equals(Scale);
+
+            public override int GetHashCode() =>
+                19 * Scale.GetHashCode()
+                + 59 * ScaledMinValue.GetHashCode()
+                + 79 * ScaledMaxValue.GetHashCode();
+        }
+
+        internal const int MaxSupportedScale = 38;
+
+        // This is a helper value. The length of the array depends on the supported precisions and scales.
+        // The current length (39) is enough to support BigQueryNumeric and BigQueryBigNumeric.
+        internal static readonly BigInteger[] PowersOf10 = Enumerable.Range(0, 39).Select(p => BigInteger.Pow(10, p)).ToArray();
+
+        // TODO: Don't require a 0 before the decimal point.
+        // TODO: Replace with manual validation if we find this is a performance bottleneck (as it was with field name validation).
+        private static readonly Regex s_validation = new Regex(@"^-?[0-9]+\.?[0-9]*$");
+
+        private static readonly ConcurrentDictionary<Parameters, NumericType> s_defined = new ConcurrentDictionary<Parameters, NumericType>();
+
+        private readonly Parameters _parameters;
+        private readonly Lazy<NumericValue> _lazyMaxValue;
+        private readonly Lazy<NumericValue> _lazyMinValue;
+        private readonly Lazy<NumericValue> _lazyEpsilon;
+        private readonly Lazy<NumericValue> _lazyZero;
+
+        private NumericType(Parameters parameters)
+        {
+            _parameters = parameters;
+            _lazyMinValue = new Lazy<NumericValue>(() => ValueFromScaledInteger(_parameters.ScaledMinValue));
```

Comment:
Do we need laziness here? We'll only go through here a few times, but the use of the lazy values could be frequent. I suspect we'd have simpler code and better performance to just handle things directly. I may be missing something :)